File: front/web/lib/exploration.ts
/**
 * Exploration gestures — the Workbench /exploration source (AIDOS step S28).
 *
 * KRD §75/§84/§118/§132: the three exploration gestures of the idea entry-stage run OVER the S27
 * Idea lifecycle. /grill challenges an intention ABOVE the wall and ROUTES it on a closed
 * three-value verdict (sharp → grilled, skips spike ; fuzzy → spiking, the floue branch ; bad →
 * rejected, traced). /spike is the ratchet-OFF, T0, THROWAWAY zone whose writes are confined to
 * /spike (an escaping write ⇒ SPIKE_WRITE_ESCAPES_ZONE). /harvest extracts the discovered intention
 * and PROPOSES a DRAFT Truth — a kernel-delta candidate with NO frozen version and NO mirror (the
 * freeze is a separate later /goal; a direct kernel/mirror write by harvest ⇒ HARVEST_CANNOT_FREEZE).
 *
 * This module is the DECLARED TWIN of the Go package back/runtime/exploration — the SAME closed
 * verdict set, the SAME spike-confinement predicate, the SAME DRAFT-Truth proposal shape (no freeze,
 * no mirror). One semantics, no drift — so /exploration renders EXACTLY what the Go deciders compute.
 * The reproducibility mirror lib/exploration.test.ts (fast-check) pins it.
 *
 * READ-ONLY (CLAUDE.md §7 ui-completeness, the wall): /exploration PROJECTS the gesture machine and
 * the confinement/harvest verdicts; recording an Idea transition rides the idea-intake MCP (ideas
 * schema, ABOVE the wall), and a PROMOTION of a DRAFT Truth writes a kernel truth via /goal under the
 * aidos writer role — never a write from this screen. The agent DB role is SELECT-only on
 * kernel/mirrors (the wall holds).
 */

import type { Idea, Proposes, Provenance, Status } from "@/lib/ideas";

/** The closed three-value grill verdict (KRD §75/§118) — no fourth value. */
export type GrillVerdict = "sharp" | "fuzzy" | "bad";

/** The closed verdict set, in canonical order (the twin of exploration.Verdicts). */
export function verdicts(): GrillVerdict[] {
	return ["sharp", "fuzzy", "bad"];
}

/** The actionable refusal shape (mirrors blockreason.BlockReason, KRD §44.5). */
export interface BlockReason {
	code: string;
	severity: string;
	explanation: string;
	howToFix: string[];
}

/** The only zone a spiking idea may write to (KRD §84) — the twin of exploration.SpikePrefix. */
export const SPIKE_PREFIX = "/spike";

/** The canonical SPIKE_WRITE_ESCAPES_ZONE BlockReason — the twin of blockreason.For (FR prose). */
export const SPIKE_WRITE_ESCAPES_ZONE: BlockReason = {
	code: "SPIKE_WRITE_ESCAPES_ZONE",
	severity: "blocking",
	explanation:
		"Refus du confinement : pendant qu'une idée est en spiking (cliquet OFF, T0, jetable), une écriture sort de la zone /spike. Le spike est jetable et ne doit JAMAIS fuir vers /kernel ni /src ; il ne gradue pas directement vers le noyau.",
	howToFix: [
		"confine_write_to_/spike : ramenez l'écriture sous le préfixe /spike — tout ce qu'un spike produit y reste, et reste jetable.",
		"run_/harvest_to_propose_a_kernel_delta : quand l'intention découverte est nette, récoltez-la pour PROPOSER un delta-noyau DRAFT.",
		"rerun aidos check : le blocage se lève dès que l'écriture est confinée à /spike.",
	],
};

/** The canonical HARVEST_CANNOT_FREEZE BlockReason — the twin of blockreason.For (FR prose). */
export const HARVEST_CANNOT_FREEZE: BlockReason = {
	code: "HARVEST_CANNOT_FREEZE",
	severity: "blocking",
	explanation:
		"Refus du mur : /harvest tente d'écrire le noyau ou un miroir directement. Harvest PROPOSE une DRAFT Truth (un delta-noyau candidat sans gel ni miroir) ; il ne gèle JAMAIS. Le gel est un /goal ultérieur séparé : l'IA esquisse, l'humain gèle.",
	howToFix: [
		"harvest_proposes_only : /harvest s'arrête à la proposition DRAFT — aucune écriture du noyau ni d'un miroir.",
		"write_mirror_run_goal_freeze : pour figer la DRAFT Truth, ouvrez un /goal séparé qui écrit son miroir (le gel).",
		"rerun aidos check : le blocage se lève dès que harvest ne vise plus le noyau/un miroir.",
	],
};

/** The status /grill routes a draft idea to, per verdict (twin of exploration.Grill). */
export function routeVerdict(verdict: GrillVerdict): Status {
	switch (verdict) {
		case "sharp":
			return "grilled";
		case "fuzzy":
			return "spiking";
		case "bad":
			return "rejected";
	}
}

/** A single write a spiking session attempts (twin of exploration.SpikeWrite). */
export interface SpikeWrite {
	path: string;
}

/** confined: true iff the write path stays inside the /spike zone (twin of SpikeWrite.Confined). */
export function confined(w: SpikeWrite): boolean {
	return w.path === SPIKE_PREFIX || w.path.startsWith(`${SPIKE_PREFIX}/`);
}

/**
 * checkSpikeWrite — the confinement decision (twin of exploration.CheckSpikeWrite). null if the
 * write is confined to /spike, else the SPIKE_WRITE_ESCAPES_ZONE BlockReason. Pure.
 */
export function checkSpikeWrite(w: SpikeWrite): BlockReason | null {
	return confined(w) ? null : SPIKE_WRITE_ESCAPES_ZONE;
}

/** The truth schemas /harvest may never write directly (the freeze is a separate /goal). */
const TRUTH_SCHEMAS = new Set(["kernel", "mirrors", "fitness"]);

/**
 * checkHarvestWrite — the wall for harvest (twin of exploration.CheckHarvestWrite). null if the
 * target is not a truth schema, else HARVEST_CANNOT_FREEZE. Harvest proposes, it never freezes.
 */
export function checkHarvestWrite(schema: string): BlockReason | null {
	return TRUTH_SCHEMAS.has(schema) ? HARVEST_CANNOT_FREEZE : null;
}

/** The marker every harvest proposal carries (twin of exploration.KindDraftTruth). */
export const KIND_DRAFT_TRUTH = "draft-truth";

/**
 * A DRAFT-Truth proposal — a kernel-delta CANDIDATE (twin of exploration.DraftTruthProposal). It
 * carries the discovered intention but, BY TYPE, has no frozen version and no mirror: the two
 * booleans are CONSTANT false, just as ideas.Idea makes them unrepresentable.
 */
export interface DraftTruthProposal {
	kind: typeof KIND_DRAFT_TRUTH;
	ideaId: string;
	proposes: Proposes;
	intent: string;
	provenance: Provenance;
	/** Always false — a DRAFT Truth never carries a frozen version (the freeze is /goal). */
	hasFrozenVersion: false;
	/** Always false — a DRAFT Truth never carries a mirror (writing the mirror IS the freeze). */
	hasMirror: false;
}

/**
 * harvest — the /harvest gesture (twin of exploration.Harvest). It advances the idea to harvested
 * and PROPOSES a DRAFT Truth carrying the discovered intention. It writes NOTHING (the wall). The
 * proposal's hasFrozenVersion/hasMirror are constant false. discovered empty falls back to intent.
 */
export function harvest(
	idea: Idea,
	discovered: string,
): { harvested: Idea; proposal: DraftTruthProposal } {
	const intent = discovered.trim() === "" ? idea.intent : discovered.trim();
	const harvested: Idea = { ...idea, status: "harvested" };
	const proposal: DraftTruthProposal = {
		kind: KIND_DRAFT_TRUTH,
		ideaId: harvested.id,
		proposes: harvested.proposes,
		intent,
		provenance: harvested.provenance,
		hasFrozenVersion: false,
		hasMirror: false,
	};
	return { harvested, proposal };
}

File: back/runtime/exploration/exploration.go
// Package exploration is the AIDOS Runtime engine of the three §75 exploration
// gestures — /grill, /spike, /harvest — run OVER the S27 ideas lifecycle (KRD
// §75/§84/§118/§132; ADR 0023). It is the harness machinery that turns "how a
// vague idea becomes a falsifiable candidate" into a typed, traceable gesture
// machine instead of a prompt that jumps straight to the kernel.
//
// REUSE, DON'T REINVENT (CLAUDE.md §6; ADR 0023): the status transitions stay owned
// by back/kernel/ideas (Grill/Spike/Harvest/Reject) — this package CALLS them and
// never re-implements a transition or the closed status set. It adds only the
// genuinely-Runtime concerns the bare lifecycle does not carry:
//
//   - the GRILL VERDICT routing (sharp → grilled ; fuzzy → grilled then spiking ;
//     bad → rejected, traced) — KRD §75/§118/§132 ;
//   - the SPIKE-WRITE CONFINEMENT predicate (every write under /spike or refused) —
//     KRD §84/§60.x ;
//   - the DRAFT-Truth PROPOSAL shape harvest emits (a kernel-delta candidate with no
//     frozen version and no mirror) — KRD §116/§118.
//
// THE WALL (CLAUDE.md §2): this package writes NOTHING. Grill/Spike/Harvest return
// VALUES; persistence of the Idea row and the proposal rides the idea-intake MCP
// (the aidos CLI role on the `ideas` schema), never the agent. Harvest never writes
// the kernel or a mirror — it PROPOSES; the human freezes later via /goal.
//
// DETERMINISM-FIRST (CLAUDE.md §6/§8): every function here is PURE and TOTAL — no
// clock, no rng, no I/O, never panics. The same input always yields the same output;
// the reproducibility mirror exploration_property_test.go pins that.
package exploration

import (
	"fmt"
	"strings"

	"github.com/steph-frtech/aidos/back/kernel/ideas"
	"github.com/steph-frtech/aidos/back/runtime/blockreason"
)

// GrillVerdict is the closed three-value outcome of /grill on a draft idea (KRD
// §75/§118). There is no fourth verdict — the set is declared, never discovered.
type GrillVerdict string

const (
	// VerdictSharp — the intention is falsifiable NOW: it can be stated as a red
	// mirror without further probing. Routes draft → grilled, skipping /spike
	// (KRD §132: "finalement je veux un code promo" → grill (clear) → harvest).
	VerdictSharp GrillVerdict = "sharp"
	// VerdictFuzzy — the intention is not-yet-falsifiable: it needs a throwaway
	// probe before it can be a red. Routes draft → grilled → spiking (the floue
	// branch, KRD §118).
	VerdictFuzzy GrillVerdict = "fuzzy"
	// VerdictBad — a bad idea: traced and kept, never silently dropped. Routes
	// draft → rejected (KRD §118 mermaid: grill -.-> rejected).
	VerdictBad GrillVerdict = "bad"
)

// Verdicts returns the closed set of grill verdicts in canonical order. The
// Workbench renders exactly these; none is invented.
func Verdicts() []GrillVerdict {
	return []GrillVerdict{VerdictSharp, VerdictFuzzy, VerdictBad}
}

// ErrUnknownVerdict is returned by Grill when the verdict is not a member of the
// closed three-value set. Grill invents no routing for an unknown verdict.
var ErrUnknownVerdict = fmt.Errorf("exploration: unknown grill verdict")

// Grill is the /grill gesture as a Runtime routing OVER the ideas lifecycle. It
// challenges a draft idea's intention ABOVE the wall and routes it on the verdict,
// DEFERRING every status transition to the pure ideas.* functions:
//
//   - sharp → ideas.Grill(idea)                 ⇒ grilled (skips spike)
//   - fuzzy → ideas.Grill then ideas.Spike      ⇒ spiking (the floue branch)
//   - bad   → ideas.Reject(idea, reason)        ⇒ rejected (traced with the reason)
//
// It never re-implements a transition; it never writes. The verdict is recorded by
// the caller (the idea-intake MCP) as part of the provenance. reason is the traced
// rejection reason; it is used only for the bad verdict (ignored otherwise).
func Grill(idea ideas.Idea, verdict GrillVerdict, reason string) (ideas.Idea, error) {
	switch verdict {
	case VerdictSharp:
		return ideas.Grill(idea)
	case VerdictFuzzy:
		grilled, err := ideas.Grill(idea)
		if err != nil {
			return ideas.Idea{}, err
		}
		return ideas.Spike(grilled)
	case VerdictBad:
		return ideas.Reject(idea, reason)
	default:
		return ideas.Idea{}, fmt.Errorf("%w: %q", ErrUnknownVerdict, verdict)
	}
}

// SpikePrefix is the only zone a spiking idea may write to (KRD §84). A spike is
// throwaway; it must not leak into /kernel or /src. Declared, never discovered.
const SpikePrefix = "/spike"

// SpikeWrite is a single write a spiking session attempts. Its confinement is the
// non-bypassable rule: the path is UNDER /spike or the write is refused.
type SpikeWrite struct {
	// Path is the target of the write (e.g. "/spike/retry-probe.go").
	Path string `json:"path"`
}

// Confined reports whether a spike write stays inside the /spike zone. True iff the
// path is exactly "/spike" or under the "/spike/" prefix. A pure total predicate.
func (w SpikeWrite) Confined() bool {
	return w.Path == SpikePrefix || strings.HasPrefix(w.Path, SpikePrefix+"/")
}

// CheckSpikeWrite is the confinement decision: nil if the write is confined to
// /spike, else the actionable SPIKE_WRITE_ESCAPES_ZONE BlockReason. PURE and TOTAL
// — the same path always yields the same verdict. This is the single authoritative
// predicate; the spike-confinement hook DEFERS to it (never re-implements it).
func CheckSpikeWrite(w SpikeWrite) *blockreason.BlockReason {
	if w.Confined() {
		return nil
	}
	br := blockreason.For(blockreason.CodeSpikeWriteEscapesZone)
	return &br
}

// DraftTruthProposal is what /harvest produces from a spiking (or grilled) idea: a
// kernel-delta CANDIDATE (KRD §116/§118). It names what the idea proposes and carries
// the discovered intention, but BY TYPE it has no frozen version and no mirror — the
// two absences that keep it a proposal, not a truth. HasFrozenVersion and HasMirror
// are CONSTANT false (mirrored on ideas.Idea making them unrepresentable): the
// methods below are the only way to read them, and they never return true. A later
// /goal promotes it by writing its mirror = the freeze; harvest stops here.
type DraftTruthProposal struct {
	// Kind is always "draft-truth" — the marker that this is a candidate, not a
	// frozen kernel truth.
	Kind string `json:"kind"`
	// IdeaID is the harvested idea's content-addressed id — the provenance back-link
	// the future frozen truth will carry (KRD §119; OQ-S28-link).
	IdeaID string `json:"idea_id"`
	// Proposes is the layer/kind the idea would become if promoted (KRD §118).
	Proposes ideas.Proposes `json:"proposes"`
	// Intent is the discovered intention extracted from the spike (the durable
	// lesson), verbatim.
	Intent string `json:"intent"`
	// Provenance is who/what engendered the idea (KRD §119), carried forward.
	Provenance ideas.Provenance `json:"provenance"`
}

// KindDraftTruth is the marker every harvest proposal carries.
const KindDraftTruth = "draft-truth"

// HasFrozenVersion is CONSTANT false: a DRAFT-Truth proposal never carries a frozen
// version (the freeze is the separate later /goal). The method exists so the absence
// is observable and the property mirror can assert it.
func (p DraftTruthProposal) HasFrozenVersion() bool { return false }

// HasMirror is CONSTANT false: a DRAFT-Truth proposal never carries a mirror (writing
// the mirror IS the freeze = /goal). The two absences are what keep it a proposal.
func (p DraftTruthProposal) HasMirror() bool { return false }

// Harvest is the /harvest gesture: it advances the idea to harvested (DEFERRING the
// transition to ideas.Harvest) and PROPOSES a DRAFT Truth carrying the discovered
// intention. It writes NOTHING — neither the kernel nor a mirror; it returns the
// harvested idea and the proposal as VALUES the idea-intake MCP persists in the
// `ideas` schema. discovered is the durable lesson extracted from the spike; empty
// falls back to the idea's own intent (never invents one).
//
// Legal from grilled (the clear branch) or spiking (the floue branch) — Harvest
// defers that check to ideas.Harvest, which refuses an illegal transition.
func Harvest(idea ideas.Idea, discovered string) (ideas.Idea, DraftTruthProposal, error) {
	harvested, err := ideas.Harvest(idea)
	if err != nil {
		return ideas.Idea{}, DraftTruthProposal{}, err
	}
	intent := strings.TrimSpace(discovered)
	if intent == "" {
		intent = harvested.Intent
	}
	proposal := DraftTruthProposal{
		Kind:       KindDraftTruth,
		IdeaID:     harvested.ID,
		Proposes:   harvested.Proposes,
		Intent:     intent,
		Provenance: harvested.Provenance,
	}
	return harvested, proposal, nil
}

// HarvestTarget is a schema /harvest might be asked to write. The closed set of truth
// schemas a harvest may NEVER write directly (the freeze is a separate /goal).
type HarvestTarget string

// CheckHarvestWrite is the wall for harvest: nil if the target is not a truth schema,
// else the actionable HARVEST_CANNOT_FREEZE BlockReason. Harvest PROPOSES a DRAFT
// Truth; it never freezes. PURE and TOTAL. The spike-confinement hook DEFERS to it.
func CheckHarvestWrite(schema string) *blockreason.BlockReason {
	switch schema {
	case "kernel", "mirrors", "fitness":
		br := blockreason.For(blockreason.CodeHarvestCannotFreeze)
		return &br
	default:
		return nil
	}
}
